### ISSUE TITLE
indent flatTernaryExpressions is broken when conditions and values are on the same line

### ISSUE THREAD

**Issue by mourner** [eslint/eslint#8481]
* **ESLint Version:** 4.0.0-alpha.0
* default parser

```js
/*eslint indent: [2, 4, {"flatTernaryExpressions": true}] */

var foo =
    d >= 10 ? 10 :
    d >= 5 ? 5 :
    d >= 2 ? 2 :
    d >= 1 ? 1 : 0;
```

**What did you expect to happen?** No errors.

**What actually happened?**

```
   5:1  error  Expected indentation of 8 spaces but found 4  indent
   6:1  error  Expected indentation of 8 spaces but found 4  indent
   7:1  error  Expected indentation of 8 spaces but found 4  indent
```

ESLint suggest the following formatting, which doesn't make sense:

```js
var foo =
    d >= 10 ? 10 :
        d >= 5 ? 5 :
        d >= 2 ? 2 :
        d >= 1 ? 1 : 0;
```

BTW, I strongly believe that `flatTernaryExpressions` should be `true` by default. I've never in my life seen this zigzag-style ternary formatting in any projects, while the flat style is pretty much everywhere.

Thanks a lot! Apart from this nitpick, the new indent rule is really awesome! Worked great on a 40k-line project I maintain.

cc @not-an-aardvark @mysticatea

**Comment by platinumazure** [comment]
Seems like a bug to me, but I'll let @not-an-aardvark confirm. Thanks for the report!

**Comment by not-an-aardvark** [comment]
This is working as intended -- the top-level ternary expression is always offset by 1 to avoid confusion in cases like this:

```js
foo ?
bar() :
baz()
```

The intent of the option is to flatten *nested* ternary expressions (while still indenting ternary expressions that aren't inside another ternary expression). Note that if the ternary expression isn't at the start of a line, there won't be any difference anyway.

Since this option hasn't been included in a stable release yet, we can consider changing the option to make it more useful (or changing the name so it's not misleading). That said, I don't think it's a good idea to remove *all* indentation for ternary expressions.

**Comment by mourner** [comment]
@not-an-aardvark I'm really confused by your answer. Does this mean the most common ternary indentation style won't be possible to support without turning off the `indent` option completely? I find the "only indent the first ternary" rule very arbitrary and inconsistent. It would be much less surprising to either indent every nesting level in a multi-line ternary, or indent none.

**Comment by mysticatea** [comment]
I expected this is valid:

```js
/*eslint indent: [2, 4, {"flatTernaryExpressions": true}] */

var foo =
    d >= 10 ? 10 :
    d >= 5 ? 5 :
    d >= 2 ? 2 :
    d >= 1 ? 1 :
    /*else*/ 0;
```

**Comment by mysticatea** [comment]
However, yes, I expected this is valid:

```js
let a = foo
    ? bar()
    : baz()
```

The common point is `:` does not make indent, I think.

**Comment by not-an-aardvark** [comment]
> @not-an-aardvark I'm really confused by your answer. Does this mean the most common ternary indentation style won't be possible to support without turning off the `indent` option completely?

@mourner I'm hoping we can avoid that situation by changing the option behavior to more accurately match user intuition.

> The common point is `:` does not make indent, I think.

@mysticatea What should the behavior be for code like this?

```js
/*eslint indent: [2, 4, {"flatTernaryExpressions": true}] */

var foo = bar ? baz :
qux;

let a = foo ?
    bar() :
baz()
```

To me, it seems like this would result in unexpected indentation in some cases. But maybe I'm not understanding the desired behavior correctly.

**Comment by mourner** [comment]
@not-an-aardvark I'd expect all subsequent lines to have the same indented position (4), same as any other statement that you break into multiple lines.

```js
var foo = bar ? baz :
    qux;

let a = foo ?
    bar() :
    baz();

var foo = bar 
    ? baz 
    : qux;

var foo = 5 +
    2 + 
    3;
```

**Comment by not-an-aardvark** [comment]
You've mentioned that both of these should be valid:

```js
// case 1
var foo = bar ? baz :
    qux;
```

```js
// case 2
var foo =
    bar ? baz :
    qux;
```

In case 1, `qux` is indented by 4 spaces from the start of the ternary. In case 2, `qux` is not indented at all from the start of the ternary. How does the rule determine which indentation to use?

I ask because the rule is consistent in cases like this when analyzing other expressions:

```js
// case 3
var foo = {
    ok: true
};
```

```js
// case 4
var foo =
    {
        ok: true
    };
```

In both case 3 and case 4, the `ok: true` property is offset by 4 spaces from the `{`, regardless of whether the object starts on its own line.

**Comment by mysticatea** [comment]
```js
var foo = bar ? baz :
qux;
```

I think the 2nd line needs one level indentation because it's the 2nd line of the initializer expression.

```js
let a = foo ?
    bar() :
baz()
```

The 3rd line looks one level unindented. I don't think `:` should unindent from the previous line.

**Comment by mysticatea** [comment]
```js
// case 3
var foo = {
    ok: true
};

// case 4
var foo =
    {
        ok: true
    };

// case 5
var foo = {
        ok: true
    },
    bar = 3;
```

I think the case 3 is a special case.
The initializer expression body has own indentation level of the object literal, so people probably want to avoid over indentation. However, initializer expressions have one indentation basically.

**Comment by mysticatea** [comment]
I know indentation is very complex...
I will get time in the next week, the long vacation in Japan, so I will investigate this.
